cli/src/ui.rs

```diff
@@ -632,14 +621,53 @@ impl Ui {
             Some(false) => "(yN)",
             None => "(yn)",
         };
-        let default_index = default.map(|c| if c { 0 } else { 1 });
-
-        let index = self.prompt_choice(
+        self.prompt_choice_with(
             &format!("{prompt} {default_str}"),
-            &["y", "n", "yes", "no", "Yes", "No", "YES", "NO"],
-            default_index,
-        )?;
-        Ok(index % 2 == 0)
+            default.map(|v| (v, if v { "y" } else { "n" })),
+            |input| {
+                if input.eq_ignore_ascii_case("y") || input.eq_ignore_ascii_case("yes") {
+                    Ok(true)
+                } else if input.eq_ignore_ascii_case("n") || input.eq_ignore_ascii_case("no") {
+                    Ok(false)
+                } else {
+                    Err("unrecognized response")
+                }
+            },
+        )
+    }
+
+    /// Repeats the given prompt until `parse(input)` returns a value.
+    ///
+    /// If the default `(value, text)` is specified, an empty input will be
+    /// mapped to the default `value`. The default `value` will also be used in
+    /// non-interactive session. The `text` will be printed in that case.
+    pub fn prompt_choice_with<T, E: fmt::Display>(
+        &self,
+        prompt: &str,
+        default: Option<(T, &str)>,
+        mut parse: impl FnMut(&str) -> Result<T, E>,
+    ) -> io::Result<T> {
+        if !Self::can_prompt() {
+            if let Some((value, text)) = default {
+                // Choose the default automatically without waiting.
+                writeln!(self.stderr(), "{prompt}: {text}")?;
+                return Ok(value);
+            }
```

Comment:
(Optional): Perhaps add a comment in between the `}`s:

```rust
// In this case, self.prompt() below will fail
```

You could also (or instead) add another line to the docstring:

```
/// If the default is not specified, empty input will lead to the function returning an error.
```

Reply:
Updated the doc comment. FWIW, "empty input" should be ignored in that case, and non-interactive prompt will fail.